File: protocol-generator/java/src/main/java/org/entitycore/protocol/codec/Base58.java
package org.entitycore.protocol.codec;

import java.math.BigInteger;

/**
 * Base58 (Bitcoin alphabet) encode/decode, hand-rolled (dodges a Maven dep + pin).
 *
 * <p>Used for peer-id formatting/parsing (V7 §1.2 / §7.3). Leading zero bytes map to
 * a leading {@code '1'} each, per the standard Base58 convention (leading-zero
 * preserving in both directions).
 */
public final class Base58 {
    private Base58() { }

    private static final String ALPHABET =
            "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
    private static final BigInteger FIFTY_EIGHT = BigInteger.valueOf(58);
    private static final int[] INDEX = new int[128];
    static {
        java.util.Arrays.fill(INDEX, -1);
        for (int i = 0; i < ALPHABET.length(); i++) {
            INDEX[ALPHABET.charAt(i)] = i;
        }
    }

    /** Encode a byte array to a Base58 string. */
    public static String encode(byte[] octets) {
        int zeros = 0;
        while (zeros < octets.length && octets[zeros] == 0) {
            zeros++;
        }
        // Big-endian unsigned magnitude.
        BigInteger n = new BigInteger(1, octets);
        StringBuilder sb = new StringBuilder();
        while (n.signum() > 0) {
            BigInteger[] qr = n.divideAndRemainder(FIFTY_EIGHT);
            sb.append(ALPHABET.charAt(qr[1].intValue()));
            n = qr[0];
        }
        sb.reverse();
        StringBuilder out = new StringBuilder();
        for (int i = 0; i < zeros; i++) {
            out.append('1');
        }
        out.append(sb);
        return out.toString();
    }

    /** Decode a Base58 string to a byte array (leading-zero preserving). */
    public static byte[] decode(String s) {
        int ones = 0;
        while (ones < s.length() && s.charAt(ones) == '1') {
            ones++;
        }
        BigInteger n = BigInteger.ZERO;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            int d = (c < 128) ? INDEX[c] : -1;
            if (d < 0) {
                throw new IllegalArgumentException("invalid base58 char: " + c);
            }
            n = n.multiply(FIFTY_EIGHT).add(BigInteger.valueOf(d));
        }
        byte[] body = n.signum() == 0 ? new byte[0] : stripSignByte(n.toByteArray());
        byte[] out = new byte[ones + body.length];
        // leading zeros already 0; copy body after them
        System.arraycopy(body, 0, out, ones, body.length);
        return out;
    }

    /** BigInteger.toByteArray may prepend a 0x00 sign byte; strip it for magnitude. */
    private static byte[] stripSignByte(byte[] b) {
        if (b.length > 1 && b[0] == 0) {
            byte[] r = new byte[b.length - 1];
            System.arraycopy(b, 1, r, 0, r.length);
            return r;
        }
        return b;
    }
}

File: protocol-generator/kotlin/src/main/kotlin/org/entitycore/protocol/codec/Base58.kt
package org.entitycore.protocol.codec

import java.math.BigInteger

/**
 * Base58 (Bitcoin alphabet) encode/decode, hand-rolled (dodges a Gradle dep + pin;
 * profile [codec].base58_library = "hand-rolled").
 *
 * Used for peer-id formatting/parsing (V7 §1.2 / §7.3). Leading zero bytes map to a
 * leading `'1'` each, per the standard Base58 convention (leading-zero preserving in
 * both directions).
 */
object Base58 {
    private const val ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
    private val FIFTY_EIGHT = BigInteger.valueOf(58)
    private val INDEX = IntArray(128) { -1 }.also {
        for (i in ALPHABET.indices) it[ALPHABET[i].code] = i
    }

    /** Encode a byte array to a Base58 string. */
    fun encode(octets: ByteArray): String {
        var zeros = 0
        while (zeros < octets.size && octets[zeros].toInt() == 0) zeros++
        var n = BigInteger(1, octets) // big-endian unsigned magnitude
        val sb = StringBuilder()
        while (n.signum() > 0) {
            val qr = n.divideAndRemainder(FIFTY_EIGHT)
            sb.append(ALPHABET[qr[1].toInt()])
            n = qr[0]
        }
        sb.reverse()
        return buildString {
            repeat(zeros) { append('1') }
            append(sb)
        }
    }

    /** Decode a Base58 string to a byte array (leading-zero preserving). */
    fun decode(s: String): ByteArray {
        var ones = 0
        while (ones < s.length && s[ones] == '1') ones++
        var n = BigInteger.ZERO
        for (c in s) {
            val d = if (c.code < 128) INDEX[c.code] else -1
            require(d >= 0) { "invalid base58 char: $c" }
            n = n.multiply(FIFTY_EIGHT).add(BigInteger.valueOf(d.toLong()))
        }
        val body = if (n.signum() == 0) ByteArray(0) else stripSignByte(n.toByteArray())
        val out = ByteArray(ones + body.size)
        body.copyInto(out, ones)
        return out
    }

    /** BigInteger.toByteArray may prepend a 0x00 sign byte; strip it for magnitude. */
    private fun stripSignByte(b: ByteArray): ByteArray =
        if (b.size > 1 && b[0].toInt() == 0) b.copyOfRange(1, b.size) else b
}
